src/Symfony/Component/Routing/Matcher/StaticUrlMatcher.php

```diff
@@ -0,0 +1,149 @@
+use Symfony\Component\Routing\RequestContext;
+
+/**
+ * StaticUrlMatcher matches URL based on a set of rules dumped by StaticUrlMatcherDumper.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+abstract class StaticUrlMatcher extends UrlMatcher implements RedirectableUrlMatcherInterface
+{
+    private $dumpedRoutes = array();
+
+    public function __construct(array $dumpedRoutes, RequestContext $context)
+    {
+        $this->dumpedRoutes = $dumpedRoutes;
+        $this->context = $context;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function match($rawPathinfo)
+    {
+        $allow = array();
+        $pathinfo = rawurldecode($rawPathinfo);
+        $trimmedPathinfo = rtrim($pathinfo, '/');
+        $context = $this->context;
+        $request = $this->request;
+        $requestMethod = $context->getMethod();
+        $host = $context->getHost();
+
+        foreach ($this->dumpedRoutes[0] as list($hostRegex, $routes)) {
+            if (null === $hostRegex) {
+                $hostMatches = array();
+            } elseif (!preg_match($hostRegex, $host, $hostMatches)) {
+                continue;
+            }
+
+            $i = 1;
+            $count = \count($routes);
+            $nextRoutes = array();
+
+            while (true) {
+                for (; $i < $count; ++$i) {
+                    if (!\is_array($prefix = $routes[$i][0])) {
+                        // nested collection
+
+                        if (null !== $prefix && 0 !== strpos($pathinfo, $prefix)) {
+                            continue;
+                        }
+
+                        if ($i < $count - 1) {
+                            $nextRoutes[] = array(1 + $i, $routes);
+                        }
+
+                        $routes = $routes[$i];
+                        $count = \count($routes);
+                        $i = 0;
+                    } else {
+                        // single route
+
+                        list($conditions, $expr, $methods, $schemes, $defaults) = $routes[$i];
+                        $matches = array();
+
+                        switch ($conditions[0]) {
+                            case 'trim':
+                                if ($conditions[1] !== $trimmedPathinfo) {
+                                    continue 2;
+                                }
+                                $checkTrailingSlash = true;
+                                break;
+                            case 'path':
+                                if ($conditions[1] !== $pathinfo) {
+                                    continue 2;
+                                }
+                                $checkTrailingSlash = false;
+                                break;
+                            case 'start':
+                                if (0 !== strpos($pathinfo, $conditions[1])) {
+                                    continue 2;
+                                }
+                                // no break
```

Comment:
instead of doing a fallthrough and then changing the matched conditions by changing the index with an int cast of the type check, it would be much more readable to duplicate the `preg_match` call, passing the right condition in each case.

Reply:
updated